File: pkg/dart2cpp/sample/cpp_generated/type_promotion_test.cpp
#include "dart2cpp.h"

// 工具宏定义

ObjectPtr<Object> getValue();
ObjectPtr<Object> getObject();
ObjectPtr<Object> getData();
ObjectPtr<Object> getItem();
ObjectPtr<Object> getValue() {
  return dart_string("test");
}

ObjectPtr<Object> getObject() {
  return dart_string("sample");
}

ObjectPtr<Object> getData() {
  return dart_literal<Int>(dart_int(1), dart_int(2), dart_int(3));
}

ObjectPtr<Object> getItem() {
  return dart_int(100);
}

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    auto obj = dart_string("Hello");
if (dart_is<String>(obj)) {
auto obj_promoted = dart_cast<String>(obj);
dart_print(obj_promoted->size());
dart_print(obj_promoted->toUpperCase());
}
Any value = dart_int(42);
if (dart_is<Int>(value)) {
auto value_promoted = dart_cast<Int>(value);
dart_print(value_promoted->operator_add(dart_int(10)));
}
auto nullable = getValue();
if (!(dart_is_null(nullable))) {
if (dart_is<String>(nullable)) {
auto nullable_promoted = dart_cast<String>(nullable);
dart_print(nullable_promoted->substring(dart_int(0), dart_int(5)));
}
}
auto x = getObject();
auto y = getObject();
if (dart_is<String>(x) && dart_is<Int>(y)) {
auto x_promoted = dart_cast<String>(x);
auto y_promoted = dart_cast<Int>(y);
dart_print(x_promoted->size()->operator_add(y_promoted));
}
auto data = getData();
if (dart_is<ObjectPtr<List<Any>>>(data)) {
auto data_promoted = dart_cast<ObjectPtr<List<Any>>>(data);
dart_print(data_promoted->size());
} else {
dart_print(data->toString());
}
auto item = getItem();
if (!(dart_is<Int>(item))) {
dart_print(dart_string("Not an integer"));
} else {
auto item_promoted = dart_cast<Int>(item);
dart_print(item_promoted->abs());
}
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}

File: pkg/dart2cpp/sample/dart/type_promotion_test.dart
/// 类型提升测试
void main() {
  // 测试1: 基本的is类型检查和类型提升
  Object obj = "Hello";
  if (obj is String) {
    // 在这个分支中，obj被提升为String类型
    print(obj.length); // 可以访问String的方法
    print(obj.toUpperCase());
  }

  // 测试2: 使用dynamic类型
  dynamic value = 42;
  if (value is int) {
    print(value + 10); // value被提升为int类型
  }

  // 测试3: 嵌套的is检查
  Object? nullable = getValue();
  if (nullable != null) {
    if (nullable is String) {
      print(nullable.substring(0, 5)); // nullable被提升为String类型
    }
  }

  // 测试4: 逻辑与运算符中的类型提升
  Object x = getObject();
  Object y = getObject();

  if (x is String && y is int) {
    // x被提升为String，y被提升为int
    print(x.length + y);
  }

  // 测试5: else分支中类型不提升
  Object data = getData();
  if (data is List) {
    print(data.length); // data是List类型
  } else {
    // data在这里不是List类型，但仍然是Object
    print(data.toString());
  }

  // 测试6: 取反的is检查
  Object item = getItem();
  if (item is! int) {
    // item不是int，但类型没有提升
    print("Not an integer");
  } else {
    // item是int类型
    print(item.abs());
  }
}

Object getValue() {
  return "test";
}

Object getObject() {
  return "sample";
}

Object getData() {
  return [1, 2, 3];
}

Object getItem() {
  return 100;
}
